### PR TITLE
Abouchau/create memberships with channels

### PR DESCRIPTION
**Pull request by N-4-Nina**

### CODE REVIEW

File: backend/test/membership.e2e-spec.ts [hunk 1]
```diff
@@ -75,34 +75,31 @@ describe('MembershipController (e2e)', () => {
...
-  });
-  it('Subscribe user1 to channel', () => {
-    return request(app.getHttpServer())
-      .post('/memberships/')
-      .send({ channelId: 1, userId: 1, role: MembershipRoleType.PARTICIPANT })
+      .send({ name: 'channel1', type: ChannelType.PUBLIC, userId: 1 })
       .expect(201);
   });
+
   it('Subscribe user2 to channel', () => {
```
Comment (tmorris42) [inline backend/test/membership.e2e-spec.ts:99]:
```suggestion
  it.skip('Subscribe user2 to channel', () => {  // Cannot test this without authenticating a user to make the request first
```

File: backend/test/membership.e2e-spec.ts [hunk 2]
```diff
@@ -75,34 +75,31 @@ describe('MembershipController (e2e)', () => {
...
     return request(app.getHttpServer())
       .post('/memberships/')
-      .send({ channelId: 1, userId: 2, role: MembershipRoleType.PARTICIPANT })
+      .send({ channelId: 1, userId: 2, role: MembershipRoleType.ADMIN })
       .expect(201);
   });
```
Comment (tmorris42) [inline backend/test/membership.e2e-spec.ts:105]:
```
  it.skip('Subscribe user2 to a channel a second time', () => { // Cannot test without the previous test
```

I cannot make a suggestion on code that wasn't changed, but the following line should be changed to the above to skip the test

**Review by tmorris42 — COMMENTED** [review]
These e2e tests won't work anymore until we redo the e2e tests to use an authenticated user. I don't think this should hold up the work. I would suggest skipping the tests and addressing it at a later time (which can be done by changing the relevant `it()` to `it.skip()`)

File: backend/test/membership.e2e-spec.ts (hunk 2 above)
Comment (N-4-Nina) [reply backend/test/membership.e2e-spec.ts:105]: and skipped

File: backend/test/membership.e2e-spec.ts (hunk 1 above)
Comment (N-4-Nina) [reply backend/test/membership.e2e-spec.ts]: skipped

File: frontend/src/components/Chat/ChannelJoinDialog.vue [hunk 3]
```diff
@@ -62,7 +62,6 @@
 import axios from 'axios';
 import { defineComponent } from 'vue';
 import { ChannelDto, CreateChannelDto } from '@/common/dto/channel.dto';
-
```
Comment (fyusuf-a) [inline frontend/src/components/Chat/ChannelJoinDialog.vue:65]: For these last 3 files, maybe it would be better to configure your editor to add a `\n` at the end of your files because those are not real changes.

File: backend/src/achievements-log/entities/achievements-log.entity.ts [hunk 4]
```diff
@@ -19,13 +19,13 @@ export class AchievementsLog {
   createdAt: Date;
 
   @Exclude()
-  @ManyToOne(() => User)
+  @ManyToOne(() => User, { onDelete: 'CASCADE' })
```
Comment (fyusuf-a) [inline backend/src/achievements-log/entities/achievements-log.entity.ts:22]: This is a good change, but it seems out of scope. Keeping our PRs short guarantees a faster review time.

File: backend/src/channels/channels.service.ts [hunk 5]
```diff
@@ -55,11 +77,9 @@ export class ChannelsService {
...
-        where: {
-          type: ChannelType.DIRECT,
-          name: channel.name,
-        },
+      await this.channelsRepository.findOneByOrFail({
+        type: ChannelType.DIRECT,
+        name: channel.name,
       })
     )
       throw 'Direct channel already exists.';
```
Comment (fyusuf-a) [inline backend/src/channels/channels.service.ts:85]:
```suggestion
   try {
      await this.channelsRepository.findOneByOrFail({
        type: ChannelType.DIRECT,
        name: channel.name,
      })
      throw 'Direct channel already exists.';
    } catch {}  
```

File: backend/src/channels/channels.service.ts [hunk 6]
```diff
@@ -40,9 +52,19 @@ export class ChannelsService {
...
-    return await this.channelsRepository.save(channel);
+    this.membershipsService.create({
+      userId: userId,
+      role: role,
+      channelId: channel.id,
+    });
+    return ret;
   }
 
   async checkDirectUsers(createChannelDto: CreateChannelDto, channel: Channel) {
```
[the remark this answers is not in the document]
Comment (fyusuf-a) [reply backend/src/channels/channels.service.ts:70]: A more direct way would be to use the decorator `@Check` in the channel entity (see `friendship.entity.ts:20`)

File: backend/src/channels/channels.service.ts [hunk 7]
```diff
@@ -40,9 +52,19 @@ export class ChannelsService {
       } else {
         channel.password = undefined;
       }
+      await this.usersRepository.findOneByOrFail({
+        id: +createChannelDto.userId,
```
Comment (fyusuf-a) [inline backend/src/channels/channels.service.ts:56]:
```suggestion
        id: userId,
```

File: backend/src/channels/channels.service.ts [hunk 8]
```diff
@@ -40,9 +52,19 @@ export class ChannelsService {
...
       channel.name = createChannelDto.name;
+      ret = await this.channelsRepository.save(channel);
+      role = MembershipRoleType.OWNER;
     }
-    return await this.channelsRepository.save(channel);
+    this.membershipsService.create({
+      userId: userId,
+      role: role,
+      channelId: channel.id,
+    });
```
Comment (fyusuf-a) [inline backend/src/channels/channels.service.ts:66]: When the `channel.type` is `DIRECT`, it seems that there is no `save`-ing in the database and that the code flow makes 2 identical calls to `this.membershipsService.create`, one in the main function, one `checkDirectUsers`. Is that on purpose? If not, I would remove any call to `save` or `create` in `checkDirectUsers` or change the function name.

File: backend/src/chat.gateway.ts [hunk 9]
```diff
@@ -131,6 +134,7 @@ export class ChatGateway
 
   @SubscribeMessage('chat-send')
   async handleSend(client: Socket, payload: ChatSendDto): Promise<string> {
+    console.log('target ' + payload.channel + ' message: ' + payload.message);
```
Comment (fyusuf-a) [inline backend/src/chat.gateway.ts:137]:
```suggestion
```

File: dtos/channels/create-channel.dto.ts [hunk 10]
```diff
@@ -1,10 +1,16 @@
...
 
 export class CreateChannelDto extends PickType(ChannelDto, [
   'name',
   'type',
   'password',
   'userOneId',
   'userTwoId',
-]) {}
+]) {
+  @ApiProperty()
```
Comment (fyusuf-a) [inline dtos/channels/create-channel.dto.ts:13]:
```suggestion
  @ApiProperty({
    description: 'The id of the owner of the channel or the creator of the channel in the case of a direct channel'
  })
```

File: backend/src/channels/channels.service.ts (hunk 6 above)
Comment (fyusuf-a) [inline backend/src/channels/channels.service.ts:70]: I would verify if the two userIds are different or both undefined: see the use of `@Check` in `friendship.entity.ts:20`

Comment (fyusuf-a) [inline backend/src/channels/channels.service.ts:70]:
```suggestion
  async checkDirectUsers(createChannelDto: CreateChannelDto, channel: Channel) {
    if (!createChannelDto.userOneId || !createChannelDto.userTwoId)
      throw 'User 1 or user 2 is not defined.';
```

File: backend/src/chat.gateway.ts [hunk 11]
```diff
@@ -122,7 +122,10 @@ export class ChatGateway
         throw new WsException('Invalid Token');
       }
       this.authenticatedSockets.set(client.id, user);
-      return 'SUCCESS';
+      user.memberships.forEach((membership) =>
+        this.handleJoin(client, { channel: membership.channel.id.toString() }),
```
Comment (fyusuf-a) [inline backend/src/chat.gateway.ts:126]:
```suggestion
  client.join(membership.channel.id.toString());
```
The method handleJoin makes unnecessary database calls in this context and is more adapted when one does not know yet if the user is in the channel.

Comment (N-4-Nina) [reply backend/src/chat.gateway.ts:126]: This belongs to another PR and should ot be here, I'll remove it.

File: backend/src/channels/channels.service.ts (hunk 8 above)
Comment (N-4-Nina) [reply backend/src/channels/channels.service.ts:66]: That's right, also it only created one of the two memberships needed. Changing it.

File: backend/src/channels/channels.service.ts (hunk 5 above)
Comment (N-4-Nina) [reply backend/src/channels/channels.service.ts:85]: Why ? Don't we catch it in the controller ?

File: dtos/channels/create-channel.dto.ts (hunk 10 above)
Comment (N-4-Nina) [reply dtos/channels/create-channel.dto.ts]: applied

File: backend/src/channels/channels.service.ts (hunk 5 above)
Comment (fyusuf-a) [reply backend/src/channels/channels.service.ts:90]: If I am correct, you want to continue with the flow of your function if the direct channel does not exist. Your `if` condition will throw in this case, stopping the execution of `createDirectChannel` and `create` with no creation of a channel in the database.

Comment (N-4-Nina) [reply backend/src/channels/channels.service.ts:90]: You're absolutely correct, my bad; but given that the name is defined as Unique in the channel entity, wouldn't it be better to simply delete this check altogether ? In the new version i actually save in the checkDirectUser, renamed createDirectChannel

**Review by fyusuf-a — APPROVED** [review]

**Comment by fyusuf-a** [comment]
Fix #120
